src/soc/ibm/power9/i2c.c

```diff
@@ -159,23 +233,65 @@ int platform_i2c_transfer(unsigned int bus, struct i2c_msg *segment,
 						printk(BIOS_INFO, "I2C transfer failed (0x%16.16llx)\n", r);
 						return -1;
 					}
-					r = read_scom(STATUS_REG(bus));
+					r = read_i2c(type, status_reg);
 				}
 
-				write_scom(FIFO_REG(bus), (uint64_t) segment[i].buf[len] << 56);
+				write_i2c_byte(type, fifo_reg, segment[i].buf[len]);
 			}
 		}
 
-		r = read_scom(STATUS_REG(bus));
+		r = read_i2c(type, status_reg);
 		while ((r & CMD_COMPLETE) == 0) {
 			if (r & UNRECOVERABLE) {
-				printk(BIOS_INFO, "I2C transfer failed (0x%16.16llx)\n", r);
+				printk(BIOS_INFO, "I2C transfer failed to complete (0x%16.16llx)\n", r);
 				return -1;
 			}
-			r = read_scom(STATUS_REG(bus));
+			r = read_i2c(type, status_reg);
 		}
 
 	}
 
 	return bytes_transfered;
 }
+
+/* Defined in fsi.h */
+void fsi_i2c_init(uint8_t chips)
+{
+	uint64_t status;
+
+	/* Nothing to do if second CPU isn't present */
+	if (!(chips & 0x02))
+		return;
+
+	/* Reset I2C */
+	write_i2c(FSI_I2C, RESET_REG, 0);
+
+	/* Wait for SCL */
+	status = read_i2c(FSI_I2C, STATUS_REG);
+	while ((status & SCL) == 0) {
+		if (status & UNRECOVERABLE)
+			die("Unrecoverable I2C error while waiting for SCL: 0x%016x\n", status);
+		status = read_i2c(FSI_I2C, STATUS_REG);
+	}
+
+	/* Send STOP command */
+	write_i2c(FSI_I2C, CMD_REG, STOP);
+
+	status = read_i2c(FSI_I2C, STATUS_REG);
+	while ((status & CMD_COMPLETE) == 0) {
+		if (status & UNRECOVERABLE)
+			die("Unrecoverable I2C error on STOP: 0x%016x\n", status);
+		status = read_i2c(FSI_I2C, STATUS_REG);
+	}
+
+	if ((status & (SCL | SDA | BUSY)) != (SCL | SDA)) {
+		/*
+		 * In case status looks like 0x_____8__ (i.e., SCL is set, but
+		 * not SDA), I2C hardware is in a messed up state, which can be
+		 * remedied by sending STOP *before* reset. Move STOP command
+		 * above reset or copy it there and remove this comment, but
+		 * keep the check.
+		 */
```

Comment:
I wasn't able to figure out how Hostboot does it from logs or code. It doesn't seem to do anything special. I'll copy sending STOP, which is what I was using for HOMER/OCC tests. I guess it happens when operation starts, but never finishes, and it might be an operation done by `pdbg` as well.